sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using Azure.Core.Json;
+
+namespace Azure.Core.Dynamic
+{
+    public partial class DynamicData
+    {
+        internal class AllowList
+        {
+            public static void AssertAllowedType<T>(T value)
+            {
+                if (value == null)
+                {
+                    return;
+                }
+
+                Type type = value.GetType();
+
+                if (!IsAllowedType(type, value))
+                {
+                    throw new NotSupportedException($"Type is not currently supported: '{type}'.");
+                }
+            }
+
+            public static bool IsAllowedType<T>(Type type, T value)
```

Comment:
Since we are now checking instances, not just types, should all these methods be renamed, i.e. IsValueAllowed?